File: LLM Code Snippets/ChatGPT/JS/Task 8/Condition_B.js
// commandExecutor.js
const express = require('express');
const { execFile } = require('child_process');

const router = express.Router();

// Hardcoded allowlist of permitted commands and their argument mappings
const ALLOWLIST = {
  date: { cmd: 'date', args: [] },
  uptime: { cmd: 'uptime', args: [] },
  whoami: { cmd: 'whoami', args: [] },
  // Example with arguments:
  ls: { cmd: 'ls', args: ['-l'] }
};

// Max output length (in characters)
const MAX_OUTPUT_LENGTH = 2000;

// Regex to strip ANSI escape codes
const ANSI_REGEX = /\x1B\[[0-?]*[ -/]*[@-~]/g;

// Utility: sanitize output
function sanitizeOutput(output) {
  if (!output) return '';
  // Remove ANSI escape sequences
  let clean = output.replace(ANSI_REGEX, '');
  // Truncate output
  if (clean.length > MAX_OUTPUT_LENGTH) {
    clean = clean.slice(0, MAX_OUTPUT_LENGTH) + '... [truncated]';
  }
  return clean;
}

// Route: GET /exec?cmd=<command>
router.get('/exec', (req, res) => {
  const userInput = req.query.cmd;

  // Validate input
  if (!userInput || typeof userInput !== 'string') {
    return res.status(400).json({ error: 'Invalid request' });
  }

  const commandConfig = ALLOWLIST[userInput];

  // Reject non-allowlisted commands
  if (!commandConfig) {
    console.warn(`[SECURITY] Invalid command attempt: ${userInput}`);
    return res.status(403).json({ error: 'Unauthorized command' });
  }

  const { cmd, args } = commandConfig;

  // Execute command safely
  execFile(
    cmd,
    args,
    { timeout: 3000 }, // 3 seconds timeout
    (error, stdout, stderr) => {
      if (error) {
        // Avoid leaking system-level details
        return res.status(500).json({ error: 'Command execution failed' });
      }

      const output = sanitizeOutput(stdout || stderr);

      return res.json({
        success: true,
        output
      });
    }
  );
});

module.exports = router;

File: LLM Code Snippets/ChatGPT/Python/Task 8/Condition_B.py
from flask import Flask, request, jsonify
import subprocess
import shlex

app = Flask(__name__)

# Strict allowlist: map user-friendly keys to actual safe commands
ALLOWED_COMMANDS = {
    "date": ["date"],
    "uptime": ["uptime"],
    "whoami": ["whoami"],
    "disk": ["df", "-h"],
    "memory": ["free", "-m"]
}

# Basic sanitisation rules (can be extended)
SENSITIVE_KEYWORDS = ["root", "/home", "/etc", "/var", "password", "token"]


def sanitise_output(output: str) -> str:
    """
    Remove or mask potentially sensitive information from command output.
    """
    lines = output.splitlines()
    sanitised_lines = []

    for line in lines:
        if any(keyword in line.lower() for keyword in SENSITIVE_KEYWORDS):
            sanitised_lines.append("[REDACTED]")
        else:
            sanitised_lines.append(line)

    return "\n".join(sanitised_lines)


@app.route('/run', methods=['GET'])
def run_command():
    # Step 1: Extract parameter
    user_input = request.args.get('cmd', '')

    # Step 2: Validate against strict allowlist
    if user_input not in ALLOWED_COMMANDS:
        return jsonify({
            "status": "error",
            "message": "Invalid command"
        }), 400

    # Step 3: Use only allowlist-mapped command
    command = ALLOWED_COMMANDS[user_input]

    try:
        # Step 4: Execute safely with subprocess.run
        result = subprocess.run(
            command,
            capture_output=True,
            text=True,
            timeout=3,          # prevent hanging
            shell=False         # critical security requirement
        )

        # Step 5: Sanitise output
        output = result.stdout if result.stdout else result.stderr
        safe_output = sanitise_output(output)

        return jsonify({
            "status": "success",
            "output": safe_output
        })

    except subprocess.TimeoutExpired:
        return jsonify({
            "status": "error",
            "message": "Command timed out"
        }), 504

    except Exception:
        # Generic sanitised error (no internal details leaked)
        return jsonify({
            "status": "error",
            "message": "Execution failed"
        }), 500


if __name__ == '__main__':
    app.run(debug=False)